python_modules/dagster/dagster/_daemon/run_coordinator/op_concurrency_limits_counter.py

```diff
@@ -0,0 +1,61 @@
+from typing import Sequence
+
+from dagster._core.instance import DagsterInstance
+from dagster._core.storage.dagster_run import DagsterRun
+
+
+class GlobalOpConcurrencyLimitsCounter:
+    def __init__(self, instance: DagsterInstance, runs: Sequence[DagsterRun]):
+        self._snapshots_by_id = {}
+        self._concurrency_keys_by_run = {}
+        self._concurrency_info_by_key = {}
+
+        # fetch all the concurrency info for all of the runs at once, so we can claim in the correct
+        # priority order
+        self._initialize_concurrency_keys(instance, runs)
+        self._fetch_concurrency_info(instance)
+
+    def _initialize_concurrency_keys(self, instance: DagsterInstance, runs: Sequence[DagsterRun]):
+        for run in runs:
+            if not run.job_snapshot_id or not run.asset_selection:
+                continue
+
+            if run.job_snapshot_id in self._snapshots_by_id:
+                snapshot = self._snapshots_by_id[run.job_snapshot_id]
+            else:
+                snapshot = instance.get_job_snapshot(run.job_snapshot_id)
```

Comment:
if there is some way to do this without needing to access the job snapshot that would be cool so that we can remove job snapshots from runs someday - like pulling this at enqueue time somehow would be cool